Add user assigned identity ability to postgresqlFlexServer

p3ck opened the pull request:
This adds the same capability that the azure cli supports for assigning userAssignedIdentities to postgresql Flexible Server.

##### SUMMARY
Adds support for User Assigned Identities for postgresql Flexible Server

This adds the same support that the azure cli has.

Fixes: #1527 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
azure_rm_postgresqlflexibleserver
azure_rm_postgresqlflexibleserver_info

p3ck committed:
Add user assigned identity ability to postgresqlFlexServer

This adds the same capability that the azure cli supports for assigning
userAssignedIdentities to postgresql Flexible Server.

p3ck commented:
I'm open to comment on this PR.  I'm not sure if I need to do anything special for SystemAssigned

I see that the azure-cli returns the following if there are not user assigned identites
postgresql_flexibleservers.models.UserAssignedIdentity(type="SystemAssigned")

I was also hoping to make the update_identities more generic so that it can be used with other modules.  But currently it refernces the postgreslql sepecific model.  I have an idea to make this more generic though.

p3ck commented on plugins/module_utils/azure_rm_common.py:
```diff
@@ -947,7 +986,10 @@ def get_mgmt_svc_client(self, client_type, base_url=None, api_version=None, supp
+            try:
+                setattr(client, '_ansible_models', importlib.import_module(client_type.__module__).models)
+            except AttributeError:
+                setattr(client, '_ansible_models', importlib.import_module(client_type.__module__)._models)
```
I had to do this when I moved to the newer azure.mgmt.rdbms library.  I think we shouldn't try and import models from the client module.

The azure client does
```
from azure.mgmt.rdbms import postgresql_flexibleservers

postgresql_flexibleservers.models.UserAssignedIdentity
```

xuzhang3 replied:
This is legacy code, it would be better to import models directly if possible

p3ck replied:
Done, but I had to leave this since this method gets executed for all modules that use the get_mgmt_svc_client method that does the authentication.

p3ck commented on plugins/module_utils/azure_rm_common.py:
```diff
@@ -1187,6 +1229,11 @@ def postgresql_flexible_client(self):
+    def postgresql_flexible_models(self):
+        self.log("Getting postgresql_flexible models")
+        return PostgreSQLFlexibleModels
+
```
I did this to follow what other code bas has done.  But I'm confused why we do this and try and import the model from the client module as well.  Is this legacy?  Which is the preferred way?

xuzhang3 replied:
I don't remember when this code was introduced, it existed before I joined the project. This is legacy code. It started as a simple way to manage models and clients and the client and models are mixed together.

xuzhang3 commented on requirements-azure.txt:
```diff
@@ -27,7 +27,7 @@ azure-mgmt-nspkg==2.0.0
 azure-mgmt-redis==13.0.0
 azure-mgmt-resource==21.1.0
-azure-mgmt-rdbms==10.0.0
+azure-mgmt-rdbms==10.2.0b12
```
Why upgrade to the beta version SDK, this is unstable and may brought break changes

p3ck replied:
I agree but how else can we support this feature??  It's not in a released version yet.  The azure-cli also depends on a beta version..  See here:

https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/setup.py#L108

xuzhang3 replied:
beta version is acceptable if there is no released SDK and we need to support the new feature.

Fred-sun commented on plugins/module_utils/azure_rm_common.py:
```diff
@@ -557,6 +558,44 @@ def update_tags(self, tags):
+
+        return changed, new_identity
+
+
```
too many blank lines
```suggestion
```

p3ck replied:
done

Fred-sun commented on plugins/modules/azure_rm_postgresqlflexibleserver.py:
```diff
@@ -533,6 +568,15 @@
+    type=dict(type='str', choices=['UserAssigned', 'None'], required=True),
+    user_assigned_identities=dict(type='dict', options=user_assigned_identities_spec, default={}),
+)
 
```
expected 2 blank lines, found 1
```suggestion


```

p3ck replied:
done

Fred-sun commented on requirements-azure.txt:
```diff
@@ -27,7 +27,7 @@ azure-mgmt-nspkg==2.0.0
 azure-mgmt-redis==13.0.0
 azure-mgmt-resource==21.1.0
-azure-mgmt-rdbms==10.0.0
+azure-mgmt-rdbms==10.2.0b12
```
azure-mgmt-rdbms 10.2.0b12 depends on azure-mgmt-core<2.0.0 and >=1.3.2 . suggeste updating to lastest version(v1.4.0)

https://github.com/ansible-collections/azure/blob/b95f0e5a1c2a8eaa70424426d53b0cacb923cd8f/requirements-azure.txt#L14

p3ck replied:
done

Fred-sun reviewed (commented):
@p3ck Please help to make some minor changes. Thank you!

p3ck commented:
I moved the update_identities method into the postgresqlflexibleserver module directly.  I had hoped to reuse it with other modules but I don't think that will be possible.

I also updated the values returned for identity to be either UserAssigned if identities are set or SystemAssigned if not.  This matches with what the azure-cli command returns to users and should make it more clear that if UserAssigned is not set then SystemAssigned is.

Fred-sun commented on tests/integration/targets/azure_rm_postgresqlflexibleserver/tasks/main.yml:
```diff
@@ -150,8 +196,23 @@
+      - user_identity_2 in output.servers[0].identity.user_assigned_identities
 
-- name: Create a post gresql flexible database(check mode)
+- name: postgresql server Idenity SystemAssigned
```
All names should start with an uppercase letter
```suggestion
- name: Postgresql server Idenity SystemAssigned
```

p3ck replied:
done

p3ck commented:
This should be all set now.  I reverted back to IdentityType None since that is what the SDK defines.
